### ISSUE TITLE
Reduce intent on array doesn't appear to localize the array

### ISSUE THREAD

**Issue by npadmana** [chapel-lang/chapel#6381]
### Summary of Problem
I'm trying to use a reduce intent in a forall loop over a block-distributed array. I would have
expected the array ``x`` below to be local inside of the loop and only accessed non-locally after the 
fact, but I get a error ``cannot access remote data in local block`` in the loop. 

I'll note that if ``x`` was a simple integer, this works just fine.

### Steps to Reproduce

**Source Code:**

```chapel
use BlockDist;

const DSpace={0.. #10};
const D : domain(1) dmapped Block(DSpace) = DSpace;

var x : [0.. #5]int;
forall d in D with (+ reduce x) {
  local {
    x += 1;
  }
}

writeln(x);
```

**Comment by npadmana** [comment]
Just to add some more information here ---

```chapel
use BlockDist;

const DSpace={0.. #10};
const D : domain(1) dmapped Block(DSpace) = DSpace;

var x : [0.. #5]int;
forall d in D with (+ reduce x) {
  writeln(here.id, ":",x.locale.id);
  local {
    x += 1;
  }
}

writeln(x);
```

shows that ``x`` is on the correct locale (i.e. it is local). Removing the ``local`` block works fine. Keeping it raises the same error.

**Comment by bradcray** [comment]
@vasslitvinov : Can you take a look at this?

**Comment by vasslitvinov** [comment]
In short, while the array for the `x` shadow variable is local, that array's domain is the same as the domain of the outer variable. The latter obviously resides on Locale 0. While doing `x += 1`, something is getting at x's domain, therefore doing a non-local access.

**Comment by npadmana** [comment]
@vasslitvinov -- just want to make sure I understand here --- should the domain of x also be shadowed, or is this the correct behavior?

**Comment by vasslitvinov** [comment]
@npadmana - Nothing in our present design talks about shadowing the domain of x.
If you'd like to argue for changing that, please support your request with a compelling use case.

**Comment by npadmana** [comment]
I would have thought that, when doing a reduction over an array variable, you would not need to communicate to access the array -- mostly for performance. 

The use case I had in mind (and where I spotted this) was building a histogram from a large distributed array. But I haven't tried to write a simple timing case here --- so I'm happy to close this until/unless I show that there is a real performance hit.

**Comment by vasslitvinov** [comment]
I see, avoiding communication due to domain accesses would be a good RVF (remote value forwarding) optimization.

**Comment by bradcray** [comment]
RVF-ing const domains has already been identified as a priority for this release cycle which seems like it could help with this case (tagging @noakesmichael, @benharsh, and @mppf on this issue, whose names are associated with this item along with Vass's).  We've seen other cases where performance has suffered due to lack of such localized copies of const domains.

I think it's also a fair question as to whether shadow arrays created to implement task intents should/could create local shadow domains as well to avoid the need to rely on such an optimization.

@npadmana : You are probably already be aware, but we're trying to pursue more of a data-centric approach to asserting locality in favor of retiring the `local` block which leads to confusing semantics in cases like this (e.g., if the domain is only localized due to an optimization like RVF, it becomes difficult to define the `local` block in a manner that is semantically consistent...).

**Comment by npadmana** [comment]
@bradcray : Agreed. I've been mostly using the ``local`` block as a comment to myself to make sure I don't accidentally shoot myself in the foot by doing accidental non-local accesses, rather than as an optimization (motivated in part by a discussion with you).

But it's probably also time for me to use ``chplvis`` or the ``CommDiagnostics`` module.

**Comment by bradcray** [comment]
I think it's completely fair game to use it as a simple probe/assertion for the time being.  But it was interesting to me that this conversation sidled up against some of the reasons we've been wanting to retire it...

That said, chplvis is really nice and the learning curve is not terribly steep, so I don't want to discourage you from using it either...

**Comment by vasslitvinov** [comment]
One aspect of this issue including RVF is discussed here: #5378

**Comment by bradcray** [comment]
@e-kayrakli: Tagging you on this because I stumbled back across it today and it seems (very) vaguely related to topics we've discussed recently like remote-value forwarding constant domains across on-clauses.
